### PR TITLE
Initial Push Frame sensor model

### PR DESCRIPTION
**Pull request by jessemapel**
Working on #366 

There is a huge amount of overlap between the line scan model and this model, but for initial testing this is working as expected with my test data.

I'll add notes on the parts of the source that are new for the push frame vs the line scan.

### CODE REVIEW

File: src/UsgsAstroPushFrameSensorModel.cpp [hunk 1]
```diff
@@ -0,0 +1,2461 @@
...
+      "m_platformFlag: {} ",
+      j["m_imageIdentifier"].dump(), j["m_sensorName"].dump(),
+      j["m_nLines"].dump(), j["m_nSamples"].dump(), j["m_platformFlag"].dump())
+
+  m_exposureDuration = j["m_exposureDuration"].get<double>();
+  m_interframeDelay = j["m_interframeDelay"].get<double>();
+
+  m_frameletHeight = j["m_frameletHeight"].get<int>();
+  m_frameletsFlipped = j["m_frameletsFlipped"].get<bool>();
+  m_frameletsOrderReversed = j["m_frameletsOrderReversed"].get<bool>();
```
Comment (jessemapel) [inline src/UsgsAstroPushFrameSensorModel.cpp:135]: New state values for push frame

File: src/UsgsAstroPushFrameSensorModel.cpp [hunk 2]
```diff
@@ -0,0 +1,2461 @@
...
+
+
+  state["m_frameletHeight"] = m_frameletHeight;
+  state["m_frameletsFlipped"] = m_frameletsFlipped;
+  state["m_frameletsOrderReversed"] = m_frameletsOrderReversed;
+  MESSAGE_LOG(
+      "m_frameletHeight: {} "
+      "m_frameletsFlipped: {} "
+      "m_frameletsOrderReversed: {}",
+      m_frameletHeight, m_frameletsFlipped, m_frameletsOrderReversed)
```
Comment (jessemapel) [inline src/UsgsAstroPushFrameSensorModel.cpp:333]: more new state parameters

File: src/UsgsAstroPushFrameSensorModel.cpp [hunk 3]
```diff
@@ -0,0 +1,2461 @@
...
+                / m_detectorLineSumming;
+  imagePt.samp = (detectorSample + m_detectorSampleOrigin - m_startingDetectorSample)
+               / m_detectorSampleSumming;
+
+  if (achievedPrecision) {
+    *achievedPrecision = desiredPrecision;
+  }
+
+  return imagePt;
+}
```
Comment (jessemapel) [inline src/UsgsAstroPushFrameSensorModel.cpp:721]: This whole function is new

File: src/UsgsAstroPushFrameSensorModel.cpp [hunk 4]
```diff
@@ -0,0 +1,2461 @@
...
+
+//***************************************************************************
+// UsgsAstroPushFrameSensorModel::groundToImage (internal version)
+//***************************************************************************
+csm::ImageCoord UsgsAstroPushFrameSensorModel::groundToImage(
+    const csm::EcefCoord& groundPt, const std::vector<double>& adj,
+    double desiredPrecision, double* achievedPrecision,
+    csm::WarningList* warnings) const {
+
+  // Binary search for the framelet that is closest to the given ground point.
```
Comment (jessemapel) [inline src/UsgsAstroPushFrameSensorModel.cpp:670]: We can probably do better than a binary search, but the search space is small enough (at most 100,000 framelets) that it converges fast enough.

File: src/UsgsAstroPushFrameSensorModel.cpp [hunk 5]
```diff
@@ -0,0 +1,2461 @@
...
+
+  double time = m_startingEphemerisTime + 0.5 * m_exposureDuration
+              + frameletNumber * m_interframeDelay;
+
+  time -= m_centerEphemerisTime;
+
+  MESSAGE_LOG("getImageTime for image line {} is {}", image_pt.line, time)
+
+  return time;
+}
```
Comment (jessemapel) [inline src/UsgsAstroPushFrameSensorModel.cpp:1224]: This is updated

File: src/UsgsAstroPushFrameSensorModel.cpp [hunk 6]
```diff
@@ -0,0 +1,2461 @@
...
+
+  double time = getImageTime(csm::ImageCoord(line, sample));
+  getAdjSensorPosVel(time, adj, xc, yc, zc, vx, vy, vz);
+
+  // Compute the line within the framelet
+  int summedFrameletHeight = m_frameletHeight / m_detectorLineSumming;
+  double frameletLine = std::fmod(line, summedFrameletHeight);
+  if (m_frameletsFlipped) {
+    frameletLine = summedFrameletHeight - frameletLine;
+  }
```
Comment (jessemapel) [inline src/UsgsAstroPushFrameSensorModel.cpp:1704]: This is the new part of this function. This covers the differences between push frame and line scan for all transformations from image space to object space.

File: src/UsgsAstroPushFrameSensorModel.cpp [hunk 7]
```diff
@@ -0,0 +1,2461 @@
...
+    throw csm::Error(csm::Error::ISD_NOT_SUPPORTED,
+                     "Could not parse the framelet geometry.",
+                     "UsgsAstroPushFrameSensorModel::constructStateFromIsd");
+  }
+  MESSAGE_LOG(
+      "m_frameletHeight: {} "
+      "m_frameletsFlipped: {} "
+      "m_frameletsOrderReversed: {} ",
+      state["m_frameletHeight"].dump(), state["m_frameletsFlipped"].dump(),
+      state["m_frameletsOrderReversed"].dump())
```
Comment (jessemapel) [inline src/UsgsAstroPushFrameSensorModel.cpp:2223]: New push frame ISD/state parameters

File: src/UsgsAstroPushFrameSensorModel.cpp [hunk 8]
```diff
@@ -0,0 +1,2461 @@
...
+
+  double xc, yc, zc, vx, vy, vz;
+  getAdjSensorPosVel(time, adj, xc, yc, zc, vx, vy, vz);
+  csm::EcefVector sensorToGround(groundPt.x - xc,
+                                 groundPt.y - yc,
+                                 groundPt.z - zc);
+  double dist = norm(sensorToGround);
+  MESSAGE_LOG("Calculated distance is {}", dist);
+  return dist;
+}
```
Comment (jessemapel) [inline src/UsgsAstroPushFrameSensorModel.cpp:2460]: New function to help with groundToImage binary search.

**Review by AustinSanders — APPROVED** [review]
I think this all looks good.  I'm interested to see what improvements can be made to the binary search

**Comment by jessemapel** [comment]
Just needs a re-review after fixing merge conflicts. The ISIS code has a bias factor that it does to optimize the binary search, but I'm not sure how much it actually helps. https://github.com/USGS-Astrogeology/ISIS3/blob/dev/isis/src/base/objs/PushFrameCameraGroundMap/PushFrameCameraGroundMap.cpp#L46

**Review by AustinSanders — APPROVED** [review]

**Pull request merged**